### ISSUE TITLE
failsafe integration tests are too slow

### ISSUE THREAD

**Issue by yegor256** [objectionary/eo#3696]
Currently, integration tests in `eo-runtime` are too slow. Let's remove them from the main build cycle (by placing them in a new `<profile/>`) and turn the on in `mvn.yml` for GitHub Actions.

Maybe even better: let's add `surefire.yml` file into the list of jobs in GHA and let `mvn.yml` run fast, without integration tests.

**Comment by yegor256** [comment]
@maxonfjvipon see also #3705

**Comment by maxonfjvipon** [comment]
@volodya-lombrozo should be faster now

**Comment by yegor256** [comment]
@maxonfjvipon when I run `mvn clean install`, I expect no integration tests to be executed, but they do run.

**Comment by yegor256** [comment]
@volodya-lombrozo according to Maven conventions, our build is as long as it can be, but you can skip integration tests on your side to make build faster: `mvn clean install -DskipITs`

**Comment by volodya-lombrozo** [comment]
@yegor256 We can add one more profile. Let's call it `long`. This way, we can place all the "long" integration tests under this profile. When we run `mvn clean install`, it will execute as quickly as possible. However, when you run `mvn clean install -Plong`, it will perform more comprehensive and thorough testing.

Using `-DskipITs` skips **all** the integration tests, but this is not what we want. We only want to skip tests that take more than one minute to run. Otherwise, we are forced to either skip **all** the tests or run them all and wait **25 minutes** for them to finish. With the `long` profile, we will have a more flexible and effective solution, in my opinion.

**Comment by yegor256** [comment]
@volodya-lombrozo this is not true:

> When we run `mvn clean install`, it will execute as quickly as possible

As quick as possible means "no tests at all", which is `-DskipUTs`.

**Comment by volodya-lombrozo** [comment]
> @volodya-lombrozo this is not true:
> 
> > When we run `mvn clean install`, it will execute as quickly as possible
> 
> As quick as possible means "no tests at all", which is `-DskipUTs`.

@yegor256 Yes, here you are right. Sorry. But I hope, you understand the idea behind by suggestion:

1. `mvn clean install -DskipITs` - very fast, insecure. I will run this command for each commit.
2. `mvn clean install` - slightly longer, more-or-less secure (what we want). I will run this command before sending a PR
3. `mvn clean install -Plong` - very long, secure. I won't run this command myself, but CI will do.

**Comment by volodya-lombrozo** [comment]
@yegor256 I would put all the tests that take more than 2 min into the `long` profile, or 1 min on you computer.

**Comment by maxonfjvipon** [comment]
@volodya-lombrozo 1 min on my computer may be X (any random number) minutes on your computer. What's the problem to use `-Dskip*` flag for you own purpose?

**Comment by volodya-lombrozo** [comment]
@maxonfjvipon Please read [this](https://github.com/objectionary/eo/issues/3696#issuecomment-2553508818) and [this](https://github.com/objectionary/eo/issues/3696#issuecomment-2553563880). I've already answered this.

**Comment by volodya-lombrozo** [comment]
@yegor256 @maxonfjvipon For example, all tests under `eo-runtime/src/test/java/integration` take more than 5 minutes on my machine:

```
INFO  c.j.m.Manifests - 62 attributes loaded from 55 stream(s) in 85ms, 62 saved, 609 ignored: ["Agent-Class", "Ant-Version", "Archiver-Version", "Automatic-Module-Name", "Bnd-LastModified", "Build-Date", "Build-Jdk", "Build-Jdk-Spec", "Build-Revision", "Build-Time", "Built-By", "Bundle-ActivationPolicy", "Bundle-Category", "Bundle-ClassPath", "Bundle-Description", "Bundle-DocURL", "Bundle-License", "Bundle-ManifestVersion", "Bundle-Name", "Bundle-NativeCode", "Bundle-RequiredExecutionEnvironment", "Bundle-SCM", "Bundle-SymbolicName", "Bundle-Vendor", "Bundle-Version", "Can-Redefine-Classes", "Can-Retransform-Classes", "Can-Set-Native-Method-Prefix", "Created-By", "Dependencies", "DynamicImport-Package", "EO-Dob", "EO-Revision", "EO-Version", "Eclipse-BuddyPolicy", "Eclipse-ExtensibleAPI", "Export-Package", "Extension-Name", "Implementation-Title", "Implementation-URL", "Implementation-Vendor", "Implementation-Vendor-Id", "Implementation-Version", "Import-Package", "Include-Resource", "JCabi-Build", "JCabi-Date", "JCabi-Version", "Main-Class", "Manifest-Version", "Multi-Release", "Originally-Created-By", "Premain-Class", "Project-Name", "Provide-Capability", "Require-Capability", "Specification-Title", "Specification-Vendor", "Specification-Version", "Tool", "X-Compile-Source-JDK", "X-Compile-Target-JDK"]
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (2s), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (6s), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (12s), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (20s), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (30s), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (42s), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (56s), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (1min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (2min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (2min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (2min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (3min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (3min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (4min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (4min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (5min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (5min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (6min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (6min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (7min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (8min), please wait...
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (8min), please wait...
```

And there are only a few examples of such tests. So we can start with them.

**Comment by volodya-lombrozo** [comment]
PS `integration.PhiUnphiIT#runsTestsAfterPhiAndUnphi` took `10 min 55 sec` on my laptop

**Comment by yegor256** [comment]
@volodya-lombrozo you suggest classifying tests by the approximate amount of time they run and then running long tests in the `long` Maven profile?

**Comment by volodya-lombrozo** [comment]
> @volodya-lombrozo you suggest classifying tests by the approximate amount of time they run and then running long tests in the `long` Maven profile?

@yegor256 Something like this.  Long story short, I suggest to move "very long test" under `long` maven profile. There is only a few examples of such tests, for example `eo-runtime/src/test/java/integration`. There are only 2 tests and they take most of the build time.

**Comment by yegor256** [comment]
@volodya-lombrozo we already classified them as "integration" tests and you can disable them with `skipITs` (assuming that they are rather long). BTW, some unit tests are also rather long (taking up to 20 seconds).

**Comment by maxonfjvipon** [comment]
@volodya-lombrozo what's the difference between using `mvn clean install -Plong` and `mvn clean install -DskipITs`?

**Comment by yegor256** [comment]
@volodya-lombrozo the principle is this: we start with one side of the "scale" -- all tests are enabled. Then, the user can disable them, one by one category: `skipITs`, `skipUTs`, `skipTests`, thus moving to another side of the "scale", where everything in disabled. 

Instead, you suggest, if I understand correctly, starting somewhere in the middle of the scale, letting user move left or right. This may be a choice, but looks less obvious to me.

**Comment by volodya-lombrozo** [comment]
@yegor256 @maxonfjvipon Ok, I checked the repository one more time and now I have completely confused. What is the difference between `skipTests`, `skipITs` and `skipUTs`  from EO point of view? 
 
**EO**

- `-DskipTests` - skip unit tests, skip integration tests. Build time is 02:03 min. 
- `-DskipITs` - run unit tests, skip integration tests: 06:11 min.   
- `-DskipUTs`  (Maven doesn't have this option, thus, I guess you invented it) - skip unit tests, but run integration tests 19:40 min 

Did I understand it correctly? Because, if it's so, it means you changed the default Maven conventions:

**Maven**

- `-DskipTests -DskipITs` - skip unit tests, skip integration tests
- `-DskipTests` - skip unit tests, but run integration tests
- `-DskipITs` - run unit tests, skip integration tests

**Comment by volodya-lombrozo** [comment]
> @volodya-lombrozo the principle is this: we start with one side of the "scale" -- all tests are enabled. Then, the user can disable them, one by one category: `skipITs`, `skipUTs`, `skipTests`, thus moving to another side of the "scale", where everything is disabled.
> 
> Instead, you suggest, if I understand correctly, starting somewhere in the middle of the scale, then letting user move left or right. This may be a possible design choice, but looks less obvious to me.

@yegor256 I suggest splitting the integration tests into two categories: 'fast' and 'long' integration tests. Since integration testing takes 19 minutes on my machine, I’m unlikely to run them. However, this approach puts me at risk of missing some bugs. Therefore, I’m asking for advice on how I can run at least some integration tests that would take less time. For example, if only 3 integration tests take 15 minutes while 100 tests take just 1 minute, I would run those 100 tests on my machine.

**Comment by yegor256** [comment]
@maxonfjvipon currently, `-DskipUTs` doesn't skip `eo-maven-plugin` tests, but it should

**Comment by yegor256** [comment]
@volodya-lombrozo this is the idea (this is how it should be):

* `-DskipTests` - skip unit tests, skip integration tests (fastest option)
* `-DskipITs` - run unit tests, skip integration tests
* `-DskipUTs - skips unit tests, runs integration tests

**Comment by volodya-lombrozo** [comment]
@yegor256 Please, run this from the project root:

```
cd eo-maven-repository
mvn clean install -DskipTests
```

**Comment by maxonfjvipon** [comment]
@yegor256 now `-DskipTests` does not skip `maven-invoker-plugin`, maybe it's a bug. I think [this](https://github.com/objectionary/eo/issues/3696#issuecomment-2554040718) is what we should achieve

**Comment by yegor256** [comment]
@maxonfjvipon I reported this: https://github.com/objectionary/eo/issues/3711 Aside from invoker, all tests are skipped when `-DskipTests` is used.

**Comment by volodya-lombrozo** [comment]
> @volodya-lombrozo this is not correct:
> 
> > -DskipTests - skip unit tests, but run integration tests
> 
> Read this: https://maven.apache.org/plugins-archives/maven-surefire-plugin-2.12.4/examples/skipping-test.html

@yegor256 Seem you are right. Thank you!

**Comment by volodya-lombrozo** [comment]
@yegor256 Is [it](https://github.com/objectionary/eo/issues/3696#issuecomment-2553983032) possible?

> @yegor256 I suggest splitting the integration tests into two categories: 'fast' and 'long' integration tests. Since integration testing takes 19 minutes on my machine, I’m unlikely to run them. However, this approach puts me at risk of missing some bugs. Therefore, I’m asking for advice on how I can run at least some integration tests that would take less time. For example, if only 3 integration tests take 15 minutes while 100 tests take just 1 minute, I would run those 100 tests on my machine.

**Comment by yegor256** [comment]
@volodya-lombrozo currently, you can do this:

First (fast):

```
mvn clean install -DskipITs
```

Then (slower):

```
mvn clean install -pl :eo-runtime -Dtest=integration.SnippetIT -DskipUTs
```

Unfortunately, sice `SnippetIT` is a parametrized test, it executes all 9 (!) scenarios and it's impossible to disable either of them. Maybe we can make this configurable via JUnit somehow. I didn't find this option.

**Comment by volodya-lombrozo** [comment]
@yegor256 @maxonfjvipon  I've tried excluding `SnippetIT`, but it didn't make a significant difference:

```xml
<excludes>
  **/SnippetIT.java
</excludes>
```

```
[INFO] Total time:  15:38 min
```

This is because `PhiUnphiIT` took 10 minutes to complete:

```
WARN  c.y.MayBeSlow - We're still running runsTestsAfterPhiAndUnphi (10min), please wait...
```

If I exclude it as well:

```xml
 <excludes>
  **/SnippetIT.java
  **/PhiUnphiIT.java
</excludes>
```

Everything runs much faster. I then checked other modules and found that there are no more integration tests besides `SnippetIT.java` and `PhiUnphiIT.java`. Thus, this approach is equivalent to using `-DskipITs`.

However, if you decide to add more integration tests, I will most likely skip them and potentially miss a bug. But that's a separate story, and I assume we will discuss it in a different issue.

So, I guess, we can close this one.
